File: workspace/3_SuperSets/test/supersets/SuperSetsTest.java
package supersets;

import static junit.framework.Assert.assertEquals;

import java.util.HashSet;
import java.util.Set;

import org.junit.Test;


public class SuperSetsTest {
	
	@Test
	public void superSetsOfEmptySet() {
		assertEquals(new HashSet<Set<Integer>>(), SetUtil.superSets(new HashSet<Integer>()));
	}
	
	@Test
	public void superSetsOfSingleElement() {
		Set<Set<Integer>> expected = new HashSet<Set<Integer>>();
		expected.add(setWithElements());
		expected.add(setWithElements(1));
		assertEquals(expected, SetUtil.superSets(setWithElements(1)));
	}
	
	@Test
	public void superSetsOfSimpleSet() {
		Set<Set<Integer>> expected = new HashSet<Set<Integer>>();
		expected.add(setWithElements());
		expected.add(setWithElements(1));
		expected.add(setWithElements(2));
		expected.add(setWithElements(1, 2));
		assertEquals(expected, SetUtil.superSets(setWithElements(1, 2)));
	}
	
	@Test
	public void superSetsOfBiggerSet() {
		Set<Set<Integer>> expected = new HashSet<Set<Integer>>();
		expected.add(setWithElements());
		expected.add(setWithElements(1));
		expected.add(setWithElements(2));
		expected.add(setWithElements(3));
		expected.add(setWithElements(1, 2));
		expected.add(setWithElements(1, 3));
		expected.add(setWithElements(2, 3));
		expected.add(setWithElements(1, 2, 3));
		
		assertEquals(expected, SetUtil.superSets(setWithElements(1, 2, 3)));
	}
	@Test
	public void superSetsOfEvenBiggerSet() {
		Set<Set<Integer>> expected = new HashSet<Set<Integer>>();
		expected.add(setWithElements());
		expected.add(setWithElements(1));
		expected.add(setWithElements(2));
		expected.add(setWithElements(3));
		expected.add(setWithElements(4));
		expected.add(setWithElements(1, 2));
		expected.add(setWithElements(1, 3));
		expected.add(setWithElements(1, 4));
		expected.add(setWithElements(2, 3));
		expected.add(setWithElements(2, 4));
		expected.add(setWithElements(3, 4));
		expected.add(setWithElements(1, 2, 3));
		expected.add(setWithElements(1, 2, 4));
		expected.add(setWithElements(1, 3, 4));
		expected.add(setWithElements(2, 3, 4));
		expected.add(setWithElements(1, 2, 3, 4));
		
		assertEquals(expected, SetUtil.superSets(setWithElements(1, 2, 3, 4)));
	}
	
	private Set<Integer> setWithElements(Integer... elements) {
		Set<Integer> result = new HashSet<Integer>();
		for (Integer integer : elements) {
			result.add(integer);
		}
		return result;
	}


}

File: solution/3_SuperSets/3_SuperSets/SuperSetsTest.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using supersets;

namespace test
{
    [TestClass]
    public class SuperSetsTest
    {
        [TestMethod]
        public void superSetsOfEmptySet() {
		    setEquals(new HashSet<HashSet<int>>(), SetUtil.superSets(new HashSet<int>()));
	    }
	
	    [TestMethod]
	    public void superSetsOfSingleElement() {
            HashSet<HashSet<int>> expected = new HashSet<HashSet<int>>();
		    expected.Add(setWithElements());
		    expected.Add(setWithElements(1));
		    setEquals(expected, SetUtil.superSets(setWithElements(1)));
	    }
	
	    [TestMethod]
	    public void superSetsOfSimpleSet() {
            HashSet<HashSet<int>> expected = new HashSet<HashSet<int>>();
		    expected.Add(setWithElements());
		    expected.Add(setWithElements(1));
		    expected.Add(setWithElements(2));
		    expected.Add(setWithElements(1, 2));
            setEquals(expected, SetUtil.superSets(setWithElements(1, 2)));
	    }
	
	    [TestMethod]
	    public void superSetsOfBiggerSet() {
            HashSet<HashSet<int>> expected = new HashSet<HashSet<int>>();
		    expected.Add(setWithElements());
		    expected.Add(setWithElements(1));
		    expected.Add(setWithElements(2));
		    expected.Add(setWithElements(3));
		    expected.Add(setWithElements(1, 2));
		    expected.Add(setWithElements(1, 3));
		    expected.Add(setWithElements(2, 3));
		    expected.Add(setWithElements(1, 2, 3));

            setEquals(expected, SetUtil.superSets(setWithElements(1, 2, 3)));
	    }

	    [TestMethod]
	    public void superSetsOfEvenBiggerSet() {
            HashSet<HashSet<int>> expected = new HashSet<HashSet<int>>();
		    expected.Add(setWithElements());
		    expected.Add(setWithElements(1));
		    expected.Add(setWithElements(2));
		    expected.Add(setWithElements(3));
		    expected.Add(setWithElements(4));
		    expected.Add(setWithElements(1, 2));
		    expected.Add(setWithElements(1, 3));
		    expected.Add(setWithElements(1, 4));
		    expected.Add(setWithElements(2, 3));
		    expected.Add(setWithElements(2, 4));
		    expected.Add(setWithElements(3, 4));
		    expected.Add(setWithElements(1, 2, 3));
		    expected.Add(setWithElements(1, 2, 4));
		    expected.Add(setWithElements(1, 3, 4));
		    expected.Add(setWithElements(2, 3, 4));
		    expected.Add(setWithElements(1, 2, 3, 4));

            setEquals(expected, SetUtil.superSets(setWithElements(1, 2, 3, 4)));
	    }
	
	    private HashSet<int> setWithElements(params int[] elements) {
		    HashSet<int> result = new HashSet<int>();
		    foreach (int integer in elements) {
			    result.Add(integer);
		    }
		    return result;
	    }

        private void setEquals(HashSet<HashSet<int>> expected, HashSet<HashSet<int>> actual)
        {
            Assert.AreEqual(expected.Count, actual.Count);
            foreach (HashSet<int> set in expected)
            {
                bool inSet = false;
                foreach (HashSet<int> actualSet in actual)
                {
                   if(set.SetEquals(actualSet)){
                       inSet = true;
                   }
                }
                Assert.IsTrue(inSet);
            }
        }

    }
}
